src/viam/sdk/common/proto_convert.hpp

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include <boost/callable_traits/args.hpp>
+#include <boost/callable_traits/return_type.hpp>
+#include <boost/mp11/algorithm.hpp>
+
+#include <type_traits>
+
+#define VIAM_SDK_API_FWD_NAMESPACE_BEGIN(module_ns_name) \
+    namespace viam {                                     \
+    namespace module_ns_name {                           \
+    namespace v1 {
+
+#define VIAM_SDK_API_FWD_NAMESPACE_END \
+    }                                  \
+    }                                  \
+    }
+
+namespace viam {
+namespace sdk {
+
+namespace proto_convert_details {
+
+// This struct should be explicitly specialized with a
+//      void operator()(const SdkType&, common::v1::ApiType*) const
+// to provide API/ABI insulated proto conversion
+template <typename SdkType>
+struct to_proto;
+
+// This struct should be explicitly specialized with a
+//      SdkType operator()(const ProtoType*) const
+// to provided API/ABI insulated construction from proto
+template <typename ProtoType>
+struct from_proto;
+
+template <typename Callable>
+using ProtoArgType =
+    std::remove_pointer_t<boost::mp11::mp_back<boost::callable_traits::args_t<Callable>>>;
+
+// Implementation struct for the omni-to_proto callable defined below.
+struct to_proto_impl {
+    template <typename SdkType>
+    auto operator()(const SdkType& t) const {
+        using ProtoReturnType = ProtoArgType<to_proto<SdkType>>;
+
+        ProtoReturnType ret;
+        to_proto<SdkType>{}(t, &ret);
+
+        return ret;
+    }
+};
+
+// Implementation struct for the omni-from_proto callable defined below.
+struct from_proto_impl {
+    template <typename ProtoType>
+    auto operator()(const ProtoType& proto) const {
+        return from_proto<ProtoType>{}(&proto);
+    }
+};
+
+}  // namespace proto_convert_details
+
+namespace v2 {
+
+/// @brief Function object implementing conversion from an SDK type to an API type.
+/// This callable works for any type with a proto_convert_details::to_proto specialization as
+/// described above.
+constexpr proto_convert_details::to_proto_impl to_proto{};
+
+/// @brief Function object implementing conversion from an API type to an SDK type.
+/// This callable works for any type with a proto_convert_details::from_proto specialization as
+/// described above.
+constexpr proto_convert_details::from_proto_impl from_proto{};
```

Comment:
These are objects defined in a header (though constexpr). Are there any concerns about linkage? I realize that's vague, but my mental model for how constexpr objects in headers work is pretty weak.

Reply:
Let me look into that actually because I was modeling this off how range-v3 does it but I see now they have a macro: https://github.com/ericniebler/range-v3/blob/7e6f34b1e820fb8321346888ef0558a0ec842b8e/include/range/v3/detail/config.hpp#L568